internal/tools/networkagent/main.go

```diff
@@ -0,0 +1,350 @@
+			NetworkNamespaceID: req.NetworkNamespaceID,
+			Ipaddresses:        []*nodenetsvc.ContainerIPAddress{resultIPAddr},
+		}
+
+		return &nodenetsvc.ConfigureContainerNetworkingResponse{
+			Interfaces: []*nodenetsvc.ContainerNetworkInterface{netInterface},
+		}, nil
+	} else if req.RequestType == nodenetsvc.RequestType_Teardown {
+		eReq := &ncproxygrpc.GetEndpointsRequest{}
+		resp, err := s.client.GetEndpoints(ctx, eReq)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, endpoint := range resp.Endpoints {
+			if endpoint.Namespace == req.NetworkNamespaceID {
+				deleteEndptReq := &ncproxygrpc.DeleteEndpointRequest{
+					Name: endpoint.Name,
+				}
+				if _, err := s.client.DeleteEndpoint(ctx, deleteEndptReq); err != nil {
+					log.G(ctx).WithField("name", endpoint.Name).Warn("failed to delete endpoint")
+				}
+			}
+		}
+
+		if networkName, ok := s.containerToNetwork[req.ContainerID]; ok {
+			deleteReq := &ncproxygrpc.DeleteNetworkRequest{
+				Name: networkName,
+			}
+			if _, err := s.client.DeleteNetwork(ctx, deleteReq); err != nil {
+				log.G(ctx).WithField("name", networkName).Warn("failed to delete network")
+			}
+			delete(s.containerToNetwork, req.ContainerID)
+		}
+
+		return &nodenetsvc.ConfigureContainerNetworkingResponse{}, nil
+
+	}
+	return nil, fmt.Errorf("invalid request type %v", req.RequestType)
+}
+
+func (s *service) addHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (_ *nodenetsvc.ConfigureNetworkingResponse, err error) {
+	return s.addHNSHelper(ctx, req, containerNamespaceID)
+}
+
+func (s *service) addHNSHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (_ *nodenetsvc.ConfigureNetworkingResponse, err error) {
+	eReq := &ncproxygrpc.GetEndpointsRequest{}
+	resp, err := s.client.GetEndpoints(ctx, eReq)
+	if err != nil {
+		return nil, err
+	}
+	log.G(ctx).WithField("endpts", resp.Endpoints).Info("ConfigureNetworking addrequest")
+
+	for _, endpoint := range resp.Endpoints {
+		if endpoint.Namespace == containerNamespaceID {
+			// add endpoints that are in the namespace as NICs
+			nicID, err := guid.NewV4()
+			if err != nil {
+				return nil, fmt.Errorf("failed to create nic GUID: %s", err)
+			}
+			nsReq := &ncproxygrpc.AddNICRequest{
+				ContainerID:  req.ContainerID,
+				NicID:        nicID.String(),
+				EndpointName: endpoint.Name,
+			}
+			if _, err := s.client.AddNIC(ctx, nsReq); err != nil {
+				return nil, err
+			}
+			s.endpointToNicID[endpoint.Name] = nicID.String()
+		}
+
+	}
+
+	defer func() {
+		if err != nil {
+			_, _ = s.teardownHelper(ctx, req, containerNamespaceID)
+		}
+	}()
+
+	return &nodenetsvc.ConfigureNetworkingResponse{}, nil
+
+}
+
+func (s *service) teardownHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	return s.teardownHNSHelper(ctx, req, containerNamespaceID)
+}
```

Comment:
What's the point of this if it just bounces the input over to `teardownHNSHelper`?

Reply:
Rather, why not just have all of the callers of this call teardownHNSHelper instead (or rename teardownHNSHelper to teardownHelper, either or haha)